src/lib/operators/join_index.cpp

```diff
@@ -311,30 +376,70 @@ void JoinIndex::_fallback_nested_loop(const ChunkID index_chunk_id, const bool t
   join_index_performance_data.chunks_scanned_without_index++;
 }
 
-// join loop that joins two segments of two columns using an iterator for the probe side,
-// and an index for the index side
+// Join loop that joins a segment of the probe table using iterators with the indexed segments of a PartialHashIndex.
+template <typename ProbeIterator>
+void JoinIndex::_data_join_probe_segment_with_indexed_segments(ProbeIterator probe_iter, ProbeIterator probe_end,
+                                                               const ChunkID probe_chunk_id,
+                                                               const std::shared_ptr<PartialHashIndex>& table_index) {
+  for (; probe_iter != probe_end; ++probe_iter) {
```

Comment:
We evaluate the join mode and the predicate condition for each tuple of the probe again and again, right? I think we should do better here (maybe templating or anything).